### PR TITLE
internal/charmstore: allow resumable uploads

### PR DESCRIPTION
**Pull request by rogpeppe**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2189/

File: docs/API.md [hunk 1]
```diff
@@ -2417,12 +2417,17 @@ time.ParseDuration function (e.g. "8h", "5m3s").
...
 	UploadId string
 
-	// Expires holds when the upload id expires (encoded
-	// in RFC3339 format).
+	// Parts holds all the known parts of the upload.
+	// Parts that haven't been uploaded yet will have nil
+	// elements.
+	Parts Parts
+
+    // Expires holds when the upload id expires (encoded
```
Comment (mhilton) [inline docs/API.md]: Can you line the comment up please?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2190/

**Review by mhilton — APPROVED** [review]
LGTM - modulo comment indenting

File: docs/API.md (hunk 1 above)
Comment (rogpeppe) [reply docs/API.md]: done

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2191/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2192/

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci-gce.jujugui.org:8080/job/charmstore-merge/619

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore/2193/

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-merge

**Pull request merged**